### DISCUSSION TITLE
Predicate function signature

### DISCUSSION THREAD

**Discussion by lmores** [dedupeio/dedupe#1146] [General]
At the momento a predicate function is defined as `PredicateFunction = Callable[[Any], Iterable[str]]`.
Could we change it to `PredicateFunction = Callable[[Any], set[str]]`?

If I am not wrong (but maybe I am) we actually never care about the order of the items inside the output of a predicate.
With this change in place we cold carry out a number of optimizations, e.g. in training.py this code:

```python
for predicate in predicates:
    coverage = frozenset(
        i
        for i, (record_1, record_2) in enumerate(pairs)
        if (set(predicate(record_1)) & set(predicate(record_2, target=True)))
    )
```

would become

```python
for predicate in predicates:
    coverage = frozenset(
        i
        for i, (record_1, record_2) in enumerate(pairs)
        if (predicate(record_1) & predicate(record_2, target=True))
    )
```

_(The above snippet could be optimized even further if we just check if the intersection is non-empty instead of building it completely.)_

Let me know!

**Comment by lmores** [comment]
@fgregg any thoughts on this?

**Comment by fgregg** [comment]
i think this is a smart idea. we implicitly rely on predicates returning unique collections and it would be good to make that explicit
